x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/repositories/GetCcrRestoreFileChunkAction.java

```diff
@@ -79,24 +77,20 @@ public void onFailure(Exception e) {
                 @Override
                 protected void doRun() throws Exception {
                     int bytesRequested = request.getSize();
-                    long fileOffset = request.getOffset();
                     ByteArray array = bigArrays.newByteArray(bytesRequested, false);
                     String fileName = request.getFileName();
                     String sessionUUID = request.getSessionUUID();
                     // This is currently safe to do because calling `onResponse` will serialize the bytes to the network layer data
                     // structure on the same thread. So the bytes will be copied before the reference is released.
                     try (ReleasablePagedBytesReference reference = new ReleasablePagedBytesReference(array, bytesRequested, array)) {
                         try (CcrRestoreSourceService.Reader reader = restoreSourceService.getSessionReader(sessionUUID, fileName)) {
-                            BytesRefIterator refIterator = reference.iterator();
-                            BytesRef ref;
-                            int bytesWritten = 0;
-                            while ((ref = refIterator.next()) != null) {
-                                byte[] refBytes = ref.bytes;
-                                reader.readFileBytes(refBytes, fileOffset + bytesWritten, ref.length);
-                                bytesWritten += ref.length;
+                            int bytesRead = reader.readFileBytes(reference);
+                            if (bytesRead < bytesRequested) {
+                                listener.onFailure(new IOException("[" + bytesRead + "] bytes requested. Only [" + bytesRead +
+                                    "] could be read from file."));
+                            } else {
+                                listener.onResponse(new GetCcrRestoreFileChunkResponse(reference));
```

Comment:
it would be nice to get some kind of progress indication on the follower side ie. a chunk ordinal or so that I can ensure I read all chucks in order. Can we sent back the offset we read from in the response and assert that on the follower end?